### PR TITLE
Pika Python 3 support

### PR DESCRIPTION
**Pull request by ztane**
As one flat commit (on top of pika's Wednesday's changes).

### CODE REVIEW

**Comment by everilae** [comment]
Missing Travis yml stuff for 2.6? :)

File: pika/channel.py [hunk 1]
```diff
@@ -206,8 +208,9 @@ def basic_consume(self, consumer_callback,
         self._validate_channel_and_callback(consumer_callback)
 
         # If a consumer tag was not passed, create one
-        consumer_tag = consumer_tag or 'ctag%i.%s' % (self.channel_number,
-                                                      uuid.uuid4().get_hex())
```
Comment (ztane) [inline pika/channel.py]: `get_hex()` is the internal method for `hex` property. It was removed in Python 3

File: .travis.yml [hunk 2]
```diff
@@ -2,14 +2,18 @@ language: python
...
+    - 3.3
+    - 3.4
 before_install:
     - sudo add-apt-repository "deb http://us.archive.ubuntu.com/ubuntu/ trusty main restricted universe multiverse"
     - sudo add-apt-repository "deb http://us.archive.ubuntu.com/ubuntu/ trusty-updates main restricted universe multiverse"
     - sudo apt-get update -qq
     - sudo apt-get install libev-dev/trusty
 install:
-    - if [[ $TRAVIS_PYTHON_VERSION == '2.6' ]]; then pip install unittest2; fi
+    - if [[ $TRAVIS_PYTHON_VERSION == '2.6' ]]; then pip install unittest2 ordereddict; fi
```
Comment (ztane) [inline .travis.yml]: Add `OrderedDict` support for Python 2.6 to make the one test fully compatible with all 5 Python flavors supported.

File: pika/adapters/blocking_connection.py [hunk 3]
```diff
@@ -344,7 +345,7 @@ def _deadline_passed(self, timeout_id):
         :rtype: bool
 
         """
-        if timeout_id not in self._timeouts.keys():
+        if timeout_id not in self._timeouts:
```
Comment (ztane) [inline pika/adapters/blocking_connection.py]: This was using `keys()` needlessly for `key in dict` test.

File: pika/credentials.py [hunk 4]
```diff
@@ -57,10 +58,12 @@ def response_for(self, start):
         :rtype: tuple(str|None, str|None)
 
         """
-        if PlainCredentials.TYPE not in start.mechanisms.split():
+        if as_bytes(PlainCredentials.TYPE) not in\
+                as_bytes(start.mechanisms).split():
             return None, None
         return (PlainCredentials.TYPE,
-                '\0%s\0%s' % (self.username, self.password))
+                b'\0' + as_bytes(self.username) +
```
Comment (ztane) [inline pika/credentials.py]: no `%` formatting for `bytes` in Python 3.0-3.4 (Python 3.5 has it)

File: pika/data.py [hunk 5]
```diff
@@ -41,48 +42,57 @@ def encode_value(pieces, value):
...
-        pieces.append(struct.pack('>ci', 'I', value))
+        pieces.append(struct.pack('>ci', b'I', value))
         return 5
-    elif isinstance(value, long):
-        pieces.append(struct.pack('>cq', 'l', value))
-        return 9
     elif isinstance(value, decimal.Decimal):
         value = value.normalize()
-        if value._exp < 0:
-            decimals = -value._exp
```
Comment (ztane) [inline pika/data.py]: This way of accessing the parts was removed in Python 3.

File: pika/data.py [hunk 6]
```diff
@@ -121,67 +131,68 @@ def decode_value(encoded, offset):
     :raises: pika.exceptions.InvalidFieldTypeException
 
     """
-    kind = encoded[offset]
+    # slice to get bytes in Python 3 and str in Python 2
+    kind = encoded[offset:offset + 1]
```
Comment (ztane) [inline pika/data.py]: The `1-slice` here is to work around the difference that in Python 2 indexing a single character in `str` still yields `str`; in Python 3, indexing a single byte yields an integer, that is the value of the byte. But 1-slice of `bytes` is still `bytes` in Python 3

File: tests/unit/callback_tests.py [hunk 7]
```diff
@@ -79,9 +83,7 @@ def test_name_or_value_str(self):
 
     def test_name_or_value_unicode(self):
         value = u'Это тест значения'
-        expectation = ('\xd0\xad\xd1\x82\xd0\xbe \xd1\x82\xd0\xb5\xd1\x81\xd1'
-                       '\x82 \xd0\xb7\xd0\xbd\xd0\xb0\xd1\x87\xd0\xb5\xd0\xbd'
-                       '\xd0\xb8\xd1\x8f')
+        expectation = 'Это тест значения'
```
Comment (ztane) [inline tests/unit/callback_tests.py]: test for the canonical str which in Python 2 is an `str` in Python 3, also an `str`

File: tests/unit/data_tests.py [hunk 8]
```diff
@@ -11,42 +11,51 @@
...
+        b'\x07decimalD\x02\x00\x00\x01:'
+        b'\x0bdecimal_tooD\x00\x00\x00\x00d'
+        b'\x07dictvalF\x00\x00\x00\x0c\x03fooS\x00\x00\x00\x03bar'
+        b'\x06intvalI\x00\x00\x00\x01'
+        b'\x07longvall\x00\x00\x00\x006e&U'
+        b'\x04nullV'
+        b'\x06strvalS\x00\x00\x00\x04Test'
+        b'\x0ctimestampvalT\x00\x00\x00\x00Ec)\x92'
+        b'\x07unicodeS\x00\x00\x00\x08utf8=\xe2\x9c\x93'
+    )
```
Comment (ztane) [inline tests/unit/data_tests.py]: The encoded table contents are reordered to match the order of insertion to the OrderedDict

File: tests/unit/data_tests.py [hunk 9]
```diff
@@ -11,42 +11,51 @@
...
+        (b'longval', long(912598613)),
+        (b'null', None),
+        (b'strval', 'Test'),
+        (b'timestampval', datetime.datetime(2006, 11, 21, 16, 30, 10)),
+        (b'unicode', u'utf8=✓')
+    ])
 
-    @unittest.skipIf(platform.python_implementation() == 'PyPy',
-                     'pypy sort order issue')
     def test_encode_table(self):
```
Comment (ztane) [inline tests/unit/data_tests.py]: With the change to `OrderedDict` this method does not fail on PyPy.

**Comment by analytik** [comment]
I'm using the latest ztane/pika@flatpatch with Python 3.4.1 on Windows, and I was wondering if this was related.

I have a simple REST application that saves some data into DB and then pushes it to RabbitMQ. But ```TornadoConnection``` sends only the first messages to Rabbit, and then just queues the rest in ```Connection.outbound_buffer``` and never sends it, even if I queue hundreds of messages. (Unless I put breakpoints somewhere in ```_flush_outbound```, then and only then I get something like 

```WARNING:pika.connection:Pika: Write buffer exceeded warning threshold at 205 bytes and an estimated 24 frames behind```

and then it's sent out. However, I suspect that is a Tornado problem, as it's not officially supported on Windows.

**Comment by ztane** [comment]
Yeah, I am not too sure about it; it can very well be a problem with this patch.

**Comment by analytik** [comment]
OK, So I've tested this on Windows with TornadoConnection in
Python 2 + master
Python 2 + flatpatch
Python 3 + flatpatch

It behaves the same, except it *does* get sent eventually on all 3 combinations (I falsely reported it doesn't get sent). The first message gets sent immediately, the next ones are sent all at once after about 30 seconds.

The only difference is that Python 3 + flatpatch reports ```Write buffer exceeded warning treshold at xx bytes...``` warnings. So I assume it's just a limitation of how pika+tornado behaves on Windows.

[Sample code here](https://gist.github.com/analytik/50873c137eaea9647132). Most of it is copied from pika documentation. (Please forgive me the Borg pattern.)

**Comment by ztane** [comment]
No assumptions, could very well still be a problem with the patch itself :D

**Comment by wjps** [comment]
@analytik, if you have a moment could you try with PR#556, the behaviour you're seeing sounds very similar. Thanks.

**Comment by analytik** [comment]
@wjps - thanks! On Python 3.4.1, Windows, with flatpatch+write-starvation-fixes, both Tornado and Select adapters work as expected. No _write buffer_ warnings anymore. :)

**Comment by wjps** [comment]
Great news.

**Comment by gmr** [comment]
Looks like there's some feedback that needs to be incorporated and then a rebase to flatten again?

**Comment by bkjones** [comment]
yeah, @ztane - I ran tests straight out of your fork and all's well: 

```shell
__________________________________________________________ summary __________________________________________________________
  py26: commands succeeded
  py27: commands succeeded
  py33: commands succeeded
  py34: commands succeeded
  congratulations :)
s757-2:pika brianj$ 
```
This was against a locally running rabbitmq 3.2.4. 

A rebase would be great at this point.

**Comment by gmr** [comment]
@ztane if you can rebase off of master, I'll freeze any additional PRs until yours is merged in. Please let me know when/if you can do that so I can coordinate with other people fixing things.

**Comment by ztane** [comment]
@gmr the latest merges to the pika master broke something (did not merge cleanly); we'd try to sort them out now in the following hours.

**Comment by vitaly-krugl** [comment]
@ztane, I am encountering some issues with SelectConnection after the latest merges: some messages don't finish sending before the ioloop terminates. Which issues are you seeing? Thx.

**Comment by ztane** [comment]
@vitaly-krugl the only problem is that the merge is not successful. When the merge is successful and there is no regression to **Python 2** (everything does not need to work in Python 3 on all platforms yet) then IMHO the pull request should be merged, we also have other things to do besides trying to keep the PR up-to-date.

**Comment by vitaly-krugl** [comment]
Ok, thanks @ztane. I was just wondering whether you were seeing the same problem that I was seeing after recent commits. Best.

**Comment by vitaly-krugl** [comment]
I think that Gavin said that he would hold off on other PRs until yours is merged, so hopefully you're really close. Good luck.

**Comment by ztane** [comment]
Superseded by a new [pull request](https://github.com/pika/pika/pull/581)

**Pull request closed without merging**